Libraries/Components/Picker/AndroidDialogPickerNativeComponent.js

```diff
@@ -22,12 +22,15 @@ import type {
 } from '../../Types/CodegenTypes';
 import type {HostComponent} from '../../Renderer/shims/ReactNativeTypes';
 import type {TextStyleProp} from '../../StyleSheet/StyleSheet';
-import type {ColorValue} from '../../StyleSheet/StyleSheetTypes';
+import type {
+  ColorValue,
+  ProcessedColorValue,
+} from '../../StyleSheet/StyleSheetTypes';
 import type {ViewProps} from '../../Components/View/ViewPropTypes';
 
 type PickerItem = $ReadOnly<{|
   label: string,
-  color?: ?Int32,
+  color?: ?ProcessedColorValue,
```

Comment:
Using this type in these native component spec files will require changing the codegen and teaching it how to handle this type. You probably need to add it here, and then fix the resulting flow errors: https://github.com/facebook/react-native/blob/master/packages/react-native-codegen/src/CodegenSchema.js#L129

Reply:
Ok, I've been teaching myself about Codegen from your video at https://www.twitch.tv/videos/368123170 😄and by reading the code in `package/react-native-codegen`.

Prior to this change the Flow type `ColorValue` was defined as `null | string`.   Most color props are typed as `ColorValue` but many other Objects keys and arguments that are colors are typed in a variety of ways: `Int32`, `number`, `number | string`.   The intention types like `Int32` and `number` are to store *processed* colors (`ColorValues` that have been processed by `processColor()`) so for this change I made the new type `ProcessedColorValue` to be explicit in the code.

From what I understand the `packages/react-native-codegen/src/parsers/flow/components/proj.js` is where Flow type names get bound to the Codegen type names like `ColorPrimitive`.   Reading the code that actually emits C++ `ColorPrimitive` types get emitted as C++ objects `SharedColor` defined in `ReactCommon/fabric/graphics/platform/cxx/Color.h`.   This `SharedColor`'s ctor only accepts `int`'s, therefore `SharedColor` objects are really `ProcessColorValue` values coming from JavaScript.

Therefore I changed `packages/react-native-codegen/src/parsers/flow/components/proj.js` so that `ProcessedColorValue` types emit as `ColorPrimitive`'s.  I pushed this change -- tests pass.

What I believe is missing is that the C++ `SharedColor` now needs to be extended to accept not only an `int` but also a map/dictionary.    But after that I'm at a loss.   How do I test the emitted C++ and how does the `SharedColor` eventually get to `RCTConvert` for iOS and `ColorPropSetter` for Android?

Reply:
These c++ Fabric prop parsing questions are maybe ones @shergin can help answer

Reply:
@tom-un What is exact shape of those objects? Is this some opaque thing (generated by some other native module) or it's just something like `{light: 42, dark: 9000}`?

`ReactCommon/fabric/graphics/platform/cxx/Color.h` implements that in general, the  simplest case that we use on Android and in unit-tests; on iOS its: `ReactCommon/fabric/graphics/platform/ios/Color.h` and that thing produces real `CGColor` objects wrapped into `shared_ptr`s.

The proper implementation of that depends on which exact info is stored into that JS object and what's supported by platform.

This PR probably has all that answers, I will read it next. :)

Reply:
The exact shapes of the Objects are defined in `NativeColorValueTypes.ios.js` for iOS and `NativeColorValueTypes.android.js` for Android.
For iOS they could look like:
- `{ semantic: ['labelColor'] }`
- `{ dynamic: { light: 42, dark: 9000 } }`
- `{ dynamic: { light: { semantic: ['lightTextColor'] }, dark: { semantic: ['darkTextColor'] } } }`
For Android they could look like:
- `{ resource_paths: ['?attr/colorPrimary'] }`

On iOS the `-[RCTConvert UIColor:]` will now examine the `id` passed in do what it has previously done if `id` is a `NSArray` or `NSNumber`.   But now if it's an `NSDictionary` it will parse the 'semantic' or 'dynamic' keys and so on and call `RCTLogConvertError` as appropriate on errors that result in meaningful RedBox error messages.

On Android its similar: the new `ColorPropSetter` class in `ViewManagersPropertyCache.java` will return an intValue if its a number, or if its a `ReadableMap` its will parse out the `resource_paths` and throw `JSApplicationCausedNativeException` as appropriate on bad input which results in meaningful RedBox error messages.

For the Codegen code, I'm not sure how the C++ `SharedColor` bridges to `RCTConvert` or `ViewManagersPropertyCache` or indeed if it even does at all.

Reply:
> it will parse the 'semantic' or 'dynamic' keys and so on and call RCTLogConvertError as appropriate on errors that result in meaningful RedBox error messages.

Is this being done in JS as well (in DEV mode)?

We'd probably want to put it in JS as well to make error messages clear and help point to the specific callsite that is invalid. We don't have the ability to do that with errors thrown from native.

Reply:
Here's what an invalid input looks like on iOS:

![image](https://user-images.githubusercontent.com/30053638/73979475-19738a80-48e3-11ea-82a8-496b73e96bfa.png)

Is the suggestion to add `  if (__DEV__) {` code to the `PlatformColor()` implementation that throws if the semantic name is unknown?   If so, there could be another native method that just return true/false given a color Object.

Reply:
I guess it's true, that the whole point of this is that only native knows the allowed values of these colors. 

The problem with those screenshots is that if you have a big codebase, there is nothing actionable to track down where the error came from. **However**, I think at DEV time the new LogBox in 0.62 (I think the actual piece of it I'm looking for will only come in 0.63) will show the JS component stack for the native error...I think.

So this is fine as is.

Reply:
>  I'm not sure how the C++ SharedColor bridges to RCTConvert or ViewManagersPropertyCache or indeed if it even does at all.

Those are not connected, `SharedColor` is a Fabric specific thing, the rest is Paper-specifics.

I am fine with the approach. Seems that the native parsing routine does not need to have an additional context to parse those color values, so it can be computed on a background thread. I am okay with landing that without Fabric-specific changes. I think it can be easily implemented later (esp. on iOS) but if you have time to do that as a separate PR, I will appreciate that!

Another consideration worth mentioning is perf: Using JS objects (esp. with the unpredictable shape) is not the most efficient way to pass data to the native side. Every single check for a key is a request to JS VM. In case when all colors are semantic, it can add meaningful overhead. That's even slower in Paper than in Fabric because Paper needs to convert JSI value into `folly::dynamic` which required iterating on object keys (and that requires creating a JS array with keys).

An alternative with better performance is using arrays with predefined semantic. E.g.: If the value is a number, use it as color value. If the value is an array, look at the first item. If it's a `1`, the next four values must be numbers representing `rgba`. If it's `2`, the next element is a string with a semantic value and so on. This way we will need to do fewer type checks, allocations, and requests to VM.

Yeah, that's not as pretty as objects but it's faster. How faster? It's really hard to measure. My gut says It might add a millisecond or so to a surface full of non-trivial colors.

Reply:
@shergin :  the good thing about the `PlatformColor()` api is that the underlying Object shape can change and it won't break the public API.   So it could easily change to arrays in a future PR that implements the Fabric specific changes -- which I'm happy to do.  I assume that in the future there will be public documentation on how to develop for and test Fabric C++ changes.